### ISSUE TITLE
`freeze` does not freeze `singleton_class`

### ISSUE THREAD

**Issue by marcandre** [oracle/truffleruby#2093]
```ruby
o = Object.new
o.singleton_class # needed for bug
o.freeze
def o.foo
  42
end # => succeeds, should raise `FrozenError`
p o.foo # => 42
```

Note that calling `singleton_class` is necessary to create the bug, I presume because it forces the creation of the `singleton_class` before freezing.

```sh
$ ruby -v
truffleruby 20.3.0-dev-80339253, like ruby 2.6.6, GraalVM CE Native [x86_64-darwin]
```

**Comment by chrisseaton** [comment]
Thanks that's an interesting bug. We should be ale to spec and fix this. Does it come from some wider code you're able to share?

**Comment by marcandre** [comment]
Found when working on this: https://github.com/ruby/ostruct/pull/15/files  (had to disable a test, search for truffle)

**Comment by bjfish** [comment]
Thanks, this is now fixed at: https://github.com/oracle/truffleruby/commit/3d1e64ad921404b3175b435572e860def9f44e5e.
